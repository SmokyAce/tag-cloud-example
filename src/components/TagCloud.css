.cloud {
    display: flex;
    flex-wrap: wrap;
    justify-content: center;
    align-items: center;
    margin: auto;
    max-width: 35%;
}

.cloud a {
    flex: 0 1 auto;
    margin: 10px;
    text-decoration: none;
}

.cloud a:hover {
    color: #3e77aa;
    text-decoration-line: underline;
}

.grow {
    transition: all .4s ease-in-out;
}

.grow:hover {
    transform: scale(1.1);
    margin: 15px;
    text-shadow: 1px 1px 3px #8f969a;
}

@media screen and (max-width: 400px) {
    .cloud {
        max-width: 90%;
    }
}

@media screen and (min-width: 400px) and (max-width: 768px) {
    .cloud {
        max-width: 75%;
    }
}

@media screen and (min-width: 768px) and (max-width: 1024px) {
    .cloud {
        max-width: 50%;
    }
}
