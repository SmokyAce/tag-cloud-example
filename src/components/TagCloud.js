import React, { useState, useEffect } from 'react';
import { Link } from 'react-router-dom';
import Spinner from './Spinner';
import { getTagCloud } from '../utils/api';
import './TagCloud.css';

function TagCloud() {
  const tags = useTags();
  const width = useWindowWidth();

  function getLabel(id) {
    return tags.ids.length > 0 && tags.byIds[id].label;
  }

  function getFontSize(id) {
    const resolution = ((width / 1920) * 100) ^ 0;
    const score = tags.byIds[id].sentimentScore;
    return (score < 50 ? 0.7 : 1) + (((score * resolution) / 100) ^ 0) / 100;
  }

  if (tags.ids.length === 0) {
    return <Spinner />;
  }

  return (
    <div className='cloud'>
      {tags.ids &&
        tags.ids.map((tagId, i) => (
          <Link
            className='grow'
            to={`/home/${tagId}`}
            style={{
              fontSize: `${getFontSize(tagId)}em`,
              lineHeight: '0.8em',
              color: `#${getRandomColor()}`
            }}
            key={i}
          >
            {getLabel(tagId)}
          </Link>
        ))}
    </div>
  );
}

function useTags() {
  const [state, setState] = useState({
    ids: [],
    byIds: {}
  });
  useEffect(() => {
    getTagCloud()
      .then(tagsByIds => {
        setState({
          ...state,
          ids: Object.keys(tagsByIds),
          byIds: tagsByIds
        });
      })
      // eslint-disable-next-line
      .catch(console.log.bind(console));
  }, []);
  return state;
}

function useWindowWidth() {
  const [width, setWidth] = useState(window.innerWidth);
  useEffect(() => {
    const handleResize = () => {
      if (Math.abs(window.innerWidth - width) > 50) {
        setWidth(window.innerWidth);
      }
    };
    window.addEventListener('resize', handleResize);
    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, [width]);
  return width;
}

function getRandomColor() {
  return Math.floor(Math.random() * 16777215).toString(16);
}

export default TagCloud;
