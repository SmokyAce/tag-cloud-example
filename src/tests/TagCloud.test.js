/* eslint-env jest */
import React from 'react';
import ReactDOM from 'react-dom';
import TestRenderer from 'react-test-renderer';
import TagCloud from '../components/TagCloud';

describe('<TagCloud />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<TagCloud />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  const testRenderer = TestRenderer.create(<TagCloud />);

  it('renders correctly', () => {
    const snapshot = testRenderer.toJSON();
    expect(snapshot).toMatchSnapshot();
  });
});
