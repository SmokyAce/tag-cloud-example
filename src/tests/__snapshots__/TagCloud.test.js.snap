// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TagCloud /> renders correctly 1`] = `
<div
  id="loading"
/>
`;
